Caching writer

stefanpenner opened the pull request:
- [x] cleaner tests
- [x] hook up with BCW
- [x] extract constants 
- [x] needs more tests
  - [x] after first build, js or css change re-generates SRI accordingly in index accordingly (needs to pass)
  - [x] test that it propagates ALL files forward.
- [x] pending a BCW bug I think i found ->  https://github.com/ember-cli/broccoli-caching-writer/issues/64

jonathanKingston committed:
Updating packages to latest version and adding snyk

jonathanKingston committed:
Adding in initial cachingWriter work

stefanpenner committed:
[FIXES #7] HTML should invalidate if any potential SRI input invalidates

stefanpenner commented on index.js:
```diff
@@ -1,20 +1,39 @@
+  }
+
+  CachingWriter.call(this, nodes, {
+    cacheInclude: [
```
are only js and css files taken into account, or are all assets?

rwjblue replied:
The regexps only look for `<script>` and `<link>` elements.

stefanpenner replied:
ah ok

rwjblue commented:
Looking good so far!

stefanpenner committed:
add failing test for propagating all files (even those not in include/exclude cache)

stefanpenner committed:
add node 4, 5
npm v3

stefanpenner commented:
I realized, there is 1 important thing this will need to do. Propagate all otherFiles, even if the cache wasn't busted. 

An example would be.
`foo.png` changes, although it doesn't trigger a build, we must still copy it. For this, I will need to add a small feature to BCW. Likely an additional hook, I need to think what that might be.

stefanpenner commented:
I haven't  had any more time tonight, and tomorrow will be quite pressed with other tasks. Simplest thing i see if to opt out of caching and land this. @rwjblue when you wake up, let me know if you think of something else. I suspect.

stefanpenner committed:
ensure all files propagate forwards (even if un-related files change)

stefanpenner committed:
listing should be its own top level testing group

rwjblue commented:
Latest update looks good to me.

stefanpenner committed:
cleanup deps

stefanpenner committed:
disable snyk.

- only “vulnerability” is a devDep of a devDep, and the devDep is actually not affected.

stefanpenner commented:
Its worth noting, working on this gave me an idea to improve ergonomics of writing more advanced broccoli plugins. BCW and BF kinda feel awkward in all by the simplest scenarios.

jonathanKingston commented:
Thanks so much for all your hard work on @stefanpenner. Totally agree on the ergonomics point.
I'll look at snyk and ESLint at some point.
